src/Http/Headers/src/ContentDispositionHeaderValue.cs

```diff
@@ -622,51 +623,57 @@ private static string Encode5987(StringSegment input)
         var maxInputBytes = Encoding.UTF8.GetMaxByteCount(input.Length);
         byte[]? bufferFromPool = null;
         Span<byte> inputBytes = maxInputBytes <= MaxStackAllocSizeBytes
-            ? stackalloc byte[MaxStackAllocSizeBytes]
+            ? stackalloc byte[maxInputBytes]
             : bufferFromPool = ArrayPool<byte>.Shared.Rent(maxInputBytes);
 
+        char[]? charBufferFromPool = null;
+        Span<char> tempCharBuffer = input.Length <= MaxStackAllocSizeChars
+            ? stackalloc char[input.Length]
+            : charBufferFromPool = ArrayPool<char>.Shared.Rent(input.Length);
+
         var bytesWritten = Encoding.UTF8.GetBytes(input, inputBytes);
         inputBytes = inputBytes[..bytesWritten];
 
-        int totalBytesConsumed = 0;
-        while (totalBytesConsumed < inputBytes.Length)
+        while (inputBytes.Length > 0)
         {
-            if (Ascii.IsValid(inputBytes[totalBytesConsumed]))
+            var length = inputBytes.IndexOfAnyExcept(Rfc5987AttrChar);
+            if (length < 0)
             {
-                // This is an ASCII char. Let's handle it ourselves.
-
-                char c = (char)inputBytes[totalBytesConsumed];
-                if (!AttrChar.Contains(c))
-                {
-                    HexEscape(builder, c);
-                }
-                else
-                {
-                    builder.Append(c);
-                }
-
-                totalBytesConsumed++;
+                length = inputBytes.Length;
             }
-            else
-            {
-                // Non-ASCII, let's rely on Rune to decode it.
+            int asciiCharCount = Encoding.ASCII.GetChars(inputBytes.Slice(0, length), tempCharBuffer);
+            builder.Append(tempCharBuffer[..asciiCharCount]);
 
-                Rune.DecodeFromUtf8(inputBytes.Slice(totalBytesConsumed), out Rune r, out int bytesConsumedForRune);
-                Contract.Assert(!r.IsAscii, "We shouldn't have gotten here if the Rune is ASCII.");
+            inputBytes = inputBytes.Slice(length);
+            if (inputBytes.Length == 0)
+            {
+                break;
+            }
 
-                for (int i = 0; i < bytesConsumedForRune; i++)
-                {
-                    HexEscape(builder, (char)inputBytes[totalBytesConsumed + i]);
-                }
+            length = inputBytes.IndexOfAny(Rfc5987AttrChar);
+            if (length < 0)
+            {
+                length = inputBytes.Length;
+            }
 
-                totalBytesConsumed += bytesConsumedForRune;
+            var toHexEscape = inputBytes.Slice(0, length);
+            while (toHexEscape.Length > 0)
+            {
+                HexEscape(builder, toHexEscape[0]);
+                toHexEscape = toHexEscape.Slice(1);
```

Comment:
Can you try a for loop instead so we don't call Slice repeatedly?

Reply:
This seems to have benefited a few ns improvement.